### PR TITLE
Conversion of Multi Class Naive Bayes classifier to estimator

### PR DESCRIPTION
**Pull request by artidoro**
Ongoing work on converting the trainers to estimators (#754). This PR converts the Multi Class Naive Bayes classification trainer.

### CODE REVIEW

File: src/Microsoft.ML.StandardLearners/Standard/MultiClass/MultiClassNaiveBayesTrainer.cs [hunk 1]
```diff
@@ -43,13 +45,58 @@ public sealed class Arguments : LearnerInputBaseWithLabel
...
 
-        public MultiClassNaiveBayesTrainer(IHostEnvironment env, Arguments args)
-            : base(env, LoadName)
+        /// <summary>
+        /// Initializes a new instance of <see cref="MultiClassNaiveBayesTrainer"/>
+        /// </summary>
+        /// <param name="env">The private instance of <see cref="IHostEnvironment"/>.</param>
+        /// <param name="labelColumn">The name of the label column.</param>
+        /// <param name="featureColumn">The name of the feature column.</param>
+        public MultiClassNaiveBayesTrainer(IHostEnvironment env, string featureColumn, string labelColumn)
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.StandardLearners/Standard/MultiClass/MultiClassNaiveBayesTrainer.cs]: >MultiClassNaiveBayesTrainer [](start = 15, length = 27)

What about extension method for the Context class?

Comment (artidoro) [reply src/Microsoft.ML.StandardLearners/Standard/MultiClass/MultiClassNaiveBayesTrainer.cs]: Ok I'll add that!

File: src/Microsoft.ML.StandardLearners/Standard/MultiClass/MultiClassNaiveBayesTrainer.cs [hunk 2]
```diff
@@ -43,13 +45,58 @@ public sealed class Arguments : LearnerInputBaseWithLabel
         private static readonly TrainerInfo _info = new TrainerInfo(normalization: false, caching: false);
         public override TrainerInfo Info => _info;
 
-        public MultiClassNaiveBayesTrainer(IHostEnvironment env, Arguments args)
-            : base(env, LoadName)
+        /// <summary>
+        /// Initializes a new instance of <see cref="MultiClassNaiveBayesTrainer"/>
+        /// </summary>
+        /// <param name="env">The private instance of <see cref="IHostEnvironment"/>.</param>
```
Comment (Zruty0) [inline src/Microsoft.ML.StandardLearners/Standard/MultiClass/MultiClassNaiveBayesTrainer.cs]: >The private instance of <see cref="IHostEnvironment"/>. [](start = 30, length = 55)

this is not helping.
`The environment to use.`

File: src/Microsoft.ML.StandardLearners/Standard/MultiClass/MultiClassNaiveBayesTrainer.cs [hunk 3]
```diff
@@ -43,13 +45,58 @@ public sealed class Arguments : LearnerInputBaseWithLabel
...
+        protected override SchemaShape.Column[] GetOutputColumnsCore(SchemaShape inputSchema)
+        {
+            bool success = inputSchema.TryFindColumn(LabelColumn.Name, out var labelCol);
+            Contracts.Assert(success);
+
+            var metadata = new SchemaShape(labelCol.Metadata.Columns.Where(x => x.Name == MetadataUtils.Kinds.KeyValues)
+                .Concat(MetadataUtils.GetTrainerOutputMetadata()));
+            return new[]
+            {
+                new SchemaShape.Column(DefaultColumnNames.Score, SchemaShape.Column.VectorKind.Vector, NumberType.R4, false, new SchemaShape(MetadataForScoreColumn())),
```
Comment (Zruty0) [inline src/Microsoft.ML.StandardLearners/Standard/MultiClass/MultiClassNaiveBayesTrainer.cs]: >MetadataForScoreColumn [](start = 141, length = 22)

1) why the disparity between this and the line below, where you initialize `metadata` right in the method?
2) method names should be verbs. So `GetScoreColumnMetadata` or something

File: test/Microsoft.ML.StaticPipelineTesting/Training.cs [hunk 4]
```diff
@@ -749,5 +749,42 @@ public void FastTreeRanking()
...
+        {
+            var env = new ConsoleEnvironment(seed: 0);
+            var dataPath = GetDataPath(TestDatasets.iris.trainFilename);
+            var dataSource = new MultiFileSource(dataPath);
+
+            var ctx = new MulticlassClassificationContext(env);
+            var reader = TextLoader.CreateReader(env,
+                c => (label: c.LoadText(0), features: c.LoadFloat(1, 4)));
+
+            MultiClassNaiveBayesPredictor pred = null;
```
Comment (Ivanidzo4ka) [inline test/Microsoft.ML.StaticPipelineTesting/Training.cs]: >MultiClassNaiveBayesPredictor [](start = 12, length = 29)

Should we expose Label and Feature histograms? Otherwise, I don't see any reason why we actually expose it.

Comment (artidoro) [reply test/Microsoft.ML.StaticPipelineTesting/Training.cs]: Created two public methods that will output a copy of the featurehistogram and labelhistogram

---
In reply to: [222122921](https://github.com/dotnet/machinelearning/pull/1111#discussion_r222122921) [](ancestors = 222122921)

File: src/Microsoft.ML.StandardLearners/Standard/MultiClass/MultiClassNaiveBayesStatics.cs [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        /// <see cref="Estimator{TTupleInShape, TTupleOutShape, TTransformer}.Fit(DataView{TTupleInShape})"/> method is called on the
+        /// <see cref="Estimator{TTupleInShape, TTupleOutShape, TTransformer}"/> instance created out of this. This delegate will receive
+        /// the linear model that was trained. Note that this action cannot change the
+        /// result in any way; it is only a way for the caller to be informed about what was learnt.</param>
+        /// <returns>The set of output columns including in order the predicted per-class likelihoods (between 0 and 1, and summing up to 1), and the predicted label.</returns>
+        public static (Vector<float> score, Key<uint, TVal> predictedLabel)
+            MultiClassNaiveBayesTrainer<TVal>(this MulticlassClassificationContext.MulticlassClassificationTrainers ctx,
+            Key<uint, TVal> label,
+            Vector<float> features,
+            Action<MultiClassNaiveBayesPredictor> onFit = null)
```
Comment (artidoro) [inline src/Microsoft.ML.StandardLearners/Standard/MultiClass/MultiClassNaiveBayesStatics.cs]: >MultiClassNaiveBayesPredictor [](start = 19, length = 29)

add xml tag

File: src/Microsoft.ML.StandardLearners/Standard/MultiClass/MultiClassNaiveBayesStatics.cs [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using Microsoft.ML.Runtime;
+using Microsoft.ML.Runtime.Data;
+using Microsoft.ML.Runtime.Internal.Calibration;
+using Microsoft.ML.Runtime.Learners;
+using Microsoft.ML.StaticPipe.Runtime;
+
```
Comment (artidoro) [inline src/Microsoft.ML.StandardLearners/Standard/MultiClass/MultiClassNaiveBayesStatics.cs]: change file name to static

File: src/Microsoft.ML.StandardLearners/Standard/MultiClass/MultiClassNaiveBayesTrainer.cs [hunk 7]
```diff
@@ -170,6 +211,40 @@ private static VersionInfo GetVersionInfo()
...
+        /// <summary>
+        /// Copies the label histogram into a buffer.
+        /// </summary>
+        /// <param name="labelHistogram">A possibly reusable array, which will
+        /// be expanded as necessary to accomodate the data.</param>
+        /// <param name="labelCount">Set to the length of the resized array, which is also the number of different labels.</param>
+        public void GetLabelHistogram(ref int[] labelHistogram, out int labelCount)
+        {
+            labelCount = _labelCount;
+            Utils.EnsureSize(ref labelHistogram, _labelCount, _labelCount);
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.StandardLearners/Standard/MultiClass/MultiClassNaiveBayesTrainer.cs]: >_labelCount, _labelCount [](start = 49, length = 24)

Why repeat twice?
Why repeat twice?

We have method which accept only one value, or you deliberately want to shrink array?

Comment (artidoro) [reply src/Microsoft.ML.StandardLearners/Standard/MultiClass/MultiClassNaiveBayesTrainer.cs]: I was thinking of returning exactly the right size for the array, but I see why if we pass a bigger array it might be better not to shrink it.

---
In reply to: [222538588](https://github.com/dotnet/machinelearning/pull/1111#discussion_r222538588) [](ancestors = 222538588)
